src/test/python/test_signer.py

```diff
@@ -0,0 +1,172 @@
+    assert signer.__get_hashed_payload(b'{"name": "rattan", "a": "value"}') == \
+           signer.__get_hashed_payload(b'{"a": "value", "name": "rattan"}')
+
+
+def test_nested_payload():
+    input_json = \
+        b"""
+        {
+            "name": "rattan",
+            "a": "value",
+            "c": {
+                "za": "value",
+                "a": "hello"
+            }
+        }
+        """
+
+    sorted_json = \
+        b"""
+        {
+            "a": "value",
+            "c": {
+                "a": "hello",
+                "za": "value"
+            },
+            "name": "rattan"
+        }
+        """
+
+    assert signer.__get_hashed_payload(input_json) == \
+           signer.__get_hashed_payload(sorted_json)
+
+
+@pytest.mark.parametrize('url, expected_encoded_url', [
+    ('/identity blah/', '/identity%20blah/'),
+    ('/identity-blah/', '/identity-blah/'),
+    ('/', '/'),
+    ('/identityABC/', '/identityABC/'),
+])
+def test_uri_endoing(url, expected_encoded_url):
+    assert signer.__encode_uri(url) == expected_encoded_url
+
+
+@freeze_time()
+def test_signature_material():
+    cvt_date = datetime.utcnow().strftime(CVT_DATE_FORMAT)
+    method = "POST"
+    url = "https://delta.covata.io/v1/secrets?hello=w+orld"
+    headers = {
+        "Content-Type": "application/json",
+        "Cvt-Date": cvt_date
+    }
+    payload = \
+        b"""
+        {
+            "name": "rattan",
+            "a": "value",
+            "c": {
+                "za": "value",
+                "a": "hello"
+            }
+        }
+        """
+
+    url_parsed = urllib.parse.urlparse(url)
+    uri = signer.__encode_uri("/".join(url_parsed.path.split("/")[2:]))
+    query = url_parsed.query.replace("+", "%20")
+    canonical_headers = 'content-type:application/json\n cvt-date:' \
+                        + cvt_date
+    signed_headers = 'content-type;cvt-date'
+    hashed_payload = \
+        '758ffa295b9a475f04aef51abd60563dabb8df1988cf6d62b9298b1d5ba6b8bf'
+    expected_canonical_request = "\n".join(
+        [method, uri, query, canonical_headers, signed_headers, hashed_payload])
+    expected_content_type = 'application/json'
+    expected_cvt_date_value = cvt_date
+    materials_from_signer = \
+        signer.__get_signature_materials(method, url, headers, payload)
+
+    assert materials_from_signer.method is method
+    assert materials_from_signer.uri == uri
+    assert materials_from_signer.headers_ == headers
+    assert materials_from_signer.query_params == query
+    assert materials_from_signer.canonical_headers == canonical_headers
+    assert materials_from_signer.canonical_request in expected_canonical_request
+    assert materials_from_signer.signed_headers == signed_headers
+    assert materials_from_signer.hashed_payload == hashed_payload
+    assert materials_from_signer.headers_['Content-Type'] is \
+           expected_content_type
+    assert materials_from_signer.headers_[
```

Comment:
as above, remove this assertion, `assert materials_from_signer.headers_ == expected_headers` already verifies this